File: go/049.go
func groupAnagrams(strs []string) [][]string {
    var res [][]string
    dict := make(map[string][]string, 0)
    for _, str := range strs {
        temp := str
        strByte := []byte(str)
        sort.SliceStable(strByte, func(i, j int) bool {
            return strByte[i] < strByte[j]
        })
        dict[string(strByte)] = append(dict[string(strByte)], temp)
    }
    for _, val := range dict {
        res = append(res, val)
    }
    return res
}

File: cpp/049.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        map<string, vector<string>> dict;
        for (auto str : strs) {
            string temp = str;
            sort(str.begin(), str.end());
            if (dict.count(str))
                dict[str].push_back(temp);
            else
                dict[str] = vector<string>{temp};
        }
        for (map<string, vector<string>>::iterator iter = dict.begin(); iter != dict.end(); ++iter) {
            res.push_back(iter->second);
        }
        return res;
    }
};
